File: concurrency/char-count/counter-thread.py
import sys
import threading
try:
    import queue           # Python 3.x
except:
    import Queue as queue  # Python 2,x
    
from char_count import char_count

def char_count_wrapper(file, queue):
    queue.put((file, char_count(file)))
    
def main(files):

    q = queue.Queue()
    tt = [threading.Thread(
        target=char_count_wrapper,
        args=(f, q))
        for f in files]

    for t in tt:
        t.start()

    for _ in files:
        f,c = q.get()
        print("%6d %s" % (c, f))

    return 0

if __name__ == "__main__":

    sys.exit(main(sys.argv[1:]))

File: concurrency/char-count/counter-thread.c
#include <stdio.h>              /* printf */
#include <stdlib.h>             /* exit */
#include <unistd.h>             /* sleep */
#include <pthread.h>            /* pthread_XXX */
#include "char_count.h"

typedef struct thread_data {

  pthread_t thread;
  char *fn;
  int count;
} thread_data_t;

void print_thread_data(thread_data_t *tdp) {

  printf("%6d %s\n", tdp->count, tdp->fn);
}

void *count(void *arg) {

  thread_data_t *tdp = (thread_data_t *) arg;

  tdp->count = char_count(tdp->fn);
  pthread_exit((void *)tdp);
}

int main(int argc, char *argv[])
{ 
  int i;
  thread_data_t *threads;

  argc--;
  argv++;

  threads = (thread_data_t *) malloc(argc * sizeof(thread_data_t));
  if (threads == NULL) {

      fprintf(stderr, "malloc failed.\n");
      exit(1);    
  }
  for (i=0; i<argc; i++) {

    thread_data_t *tdp = &threads[i];

    tdp->fn = argv[i];
    if (pthread_create(&(tdp->thread), NULL, &count, tdp)) {
      
      fprintf(stderr, "pthread_create failed.\n");
      exit(1);
    }
  }
  for (i=0; i<argc; i++) {

    thread_data_t *tdp = &threads[i];

    if (pthread_join(tdp->thread, NULL)) {

      fprintf(stderr, "join on thread %d failed.\n", i);
      exit(1);
    }
    print_thread_data(tdp);
  }
  exit(0);
}
